src/Illuminate/Database/Query/JsonPath.php

```diff
@@ -0,0 +1,245 @@
+                    break;
+                case 'identifier':
+                    $carry .= self::SEPARATOR . $item['name'];
+                    break;
+                case 'cell':
+                    $carry .= '[' . $item['name'] . ']';
+                    break;
+                case 'match':
+                    $carry .= (in_array($lastType, ['cell', 'match']) ? '.' : '') . '**"' . $item['name'] . '"';
+                    break;
+            }
+
+            $lastType = $item['type'] ?? null;
+
+            return $carry;
+        }, $this->scope);
+    }
+
+    /**
+     * Add a path for an ECMAScript identifier.
+     *
+     * @param string $name
+     *
+     * @return $this
+     *
+     * @link http://www.ecma-international.org/ecma-262/5.1/#sec-7.6
+     */
+    public function identifier(string $name): self
+    {
+        return $this->addPathLeg('identifier', $name);
+    }
+
+    /**
+     * Add a match path leg for a suffix.
+     *
+     * Generates paths using the `**` token. Paths generated are like
+     * `$.foo**suffix` and `$.foo[0].**suffix`.
+     *
+     * @param string $suffix
+     *
+     * @return $this
+     */
+    public function matches(string $suffix): self
+    {
+        return $this->addPathLeg('match', $suffix);
+    }
+
+    /**
+     * Add a member path leg.
+     *
+     * Member path legs are Json object keys. In the example
+     * `$.foo[0].name` both `foo` and `name` are members.
+     *
+     * @param string $name
+     *
+     * @return static
+     */
+    public function member(string $name): self
+    {
+        // If the member is an asterisk we want to select all members.
+        if ($name === '*') {
+            return $this->allMembers();
+        }
+
+        // If the member contains a double asterisk it's a match path leg
+        // so we're going to want to pass that where appropriate.
+        if (Str::contains($name, '**')) {
+            [$prefix, $suffix] = explode('**', $name);
+
+            if (! empty($prefix)) {
+                return $this->member($prefix)->matches($suffix);
+            }
+
+            return $this->matches($suffix);
+        }
+
+        return $this->addPathLeg('member', $name);
+    }
+
+    /**
+     * Always returns false as there's no real reason to
+     * check whether an index exists.
+     *
+     * @param mixed $offset
+     *
+     * @return bool
+     */
+    public function offsetExists($offset): bool
+    {
+        return false;
+    }
+
+    /**
+     * Allows for the setting of cell index as if accessing the object as
+     * an array.
+     *
+     * @param mixed $offset
+     *
+     * @return static
+     */
+    public function offsetGet($offset)
+    {
+        return $this->cell($offset);
+    }
+
+    public function offsetSet($offset, $value): void
+    {
+        // This method is intentionally empty
+    }
+
+    public function offsetUnset($offset): void
```

Comment:
Please don't use void return types anywhere. Also, all methods need phpdoc.

Reply:
Why no void return type?

Reply:
@ollieread we don't use return types in the framework

Reply:
Understood.

Reply:
Yes

Reply:
> @ollieread we don't use return types in the framework

just curious - why? are there any specific reasons?